csharp/ql/lib/semmle/code/csharp/security/dataflow/UrlRedirectQuery.qll

```diff
@@ -163,6 +164,32 @@ class ConcatenationSanitizer extends Sanitizer {
   }
 }
 
+/**
+ * A string interpolation expression, where the first part (before any inserts) of the
+ * expression contains the character "?".
+ *
+ * This is considered a sanitizer by the same reasoning as `ConcatenationSanitizer`.
+ */
+private class InterpolationSanitizer extends Sanitizer {
+  InterpolationSanitizer() {
+    this.getExpr().(InterpolatedStringExpr).getText(0).getValue().matches("%?%")
+  }
+}
+
+/**
+ * A call to `string.Format`, where the format expression (before any inserts)
+ * contains the character "?".
+ *
+ * This is considered a sanitizer by the same reasoning as `ConcatenationSanitizer`.
+ */
+private class StringFormatSanitizer extends Sanitizer {
+  StringFormatSanitizer() {
+    exists(FormatCall c, Expr e | c = this.getExpr() and e = c.getFormatExpr() |
+      e.(StringLiteral).getValue().splitAt("{0}", 0).matches("%?%")
```

Comment:
Doesn't this assume that the inserts are always in order? I don't think it works for
```csharp
string.Format("{1}?={0}", safe, taint)
```

Reply:
Ah yes, that is right. I will fix this.